google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/LargeReadRowsResumptionStrategy.java

```diff
@@ -0,0 +1,169 @@
+import com.google.cloud.bigtable.data.v2.stub.BigtableStreamResumptionStrategy;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+import java.util.Base64;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * An implementation of a {@link StreamResumptionStrategy} for merged rows. This class tracks -
+ *
+ * <ul>
+ *   <li>row key for the last row that was read successfully
+ *   <li>row key for large-row that couldn't be read
+ *   <li>list of all row keys for large-rows
+ * </ul>
+ *
+ * Upon retry this class builds a request to omit the large rows & retry from the last row key that
+ * was successfully read.
+ *
+ * <p>This class is considered an internal implementation detail and not meant to be used by
+ * applications.
+ */
+@InternalApi
+public class LargeReadRowsResumptionStrategy<RowT>
+    extends BigtableStreamResumptionStrategy<ReadRowsRequest, RowT> {
+  private final RowAdapter<RowT> rowAdapter;
+  private ByteString lastSuccessKey = ByteString.EMPTY;
+  // Number of rows processed excluding Marker row.
+  private long numProcessed;
+  private ByteString largeRowKey = ByteString.EMPTY;
+  private final AtomicInteger largeRowsCount = new AtomicInteger(0);
+
+  private ReadRowsRequest originalRequest;
+
+  private ReadRowsRequest lastModifiedRequestOnTimeOfError;
+
+  private RowSet previousFailedRequestRowset = null;
+
+  public LargeReadRowsResumptionStrategy(RowAdapter<RowT> rowAdapter) {
+    this.rowAdapter = rowAdapter;
+  }
+
+  @Override
+  public boolean canResume() {
+    return true;
+  }
+
+  @Override
+  public StreamResumptionStrategy<ReadRowsRequest, RowT> createNew() {
+    return new LargeReadRowsResumptionStrategy<>(rowAdapter);
+  }
+
+  @Override
+  public RowT processResponse(RowT response) {
+    // Last key can come from both the last processed row key and a synthetic row marker. The
+    // synthetic row marker is emitted when the server has read a lot of data that was filtered out.
+    // The row marker can be used to trim the start of the scan, but does not contribute to the row
+    // limit.
+    lastSuccessKey = rowAdapter.getKey(response);
+
+    if (!rowAdapter.isScanMarkerRow(response)) {
+      // Only real rows count towards the rows limit.
+      numProcessed++;
+    }
+    return response;
+  }
+
+  public Throwable processError(Throwable throwable) {
+    String rowKeyExtracted = extractLargeRowKey(throwable);
+    if (rowKeyExtracted != null) {
+      this.largeRowKey = ByteString.copyFromUtf8(rowKeyExtracted);
+      largeRowsCount.addAndGet(1);
+      numProcessed = numProcessed + 1;
+    }
+    return throwable;
+  }
+
+  private String extractLargeRowKey(Throwable t) {
+    if (t instanceof ApiException
+        && ((ApiException) t).getReason() != null
+        && ((ApiException) t).getReason().equals("LargeRowReadError")) {
+      String rowKey = ((ApiException) t).getMetadata().get("rowKey");
```

Comment:
this is updated in the latest cl. the key is "rowKeyBase64Encoded". And we should decode it.

Reply:
```suggestion
      String rowKey = ((ApiException) t).getMetadata().get("rowKeyBase64Encoded");
```